src/highdicom/seg/sop.py

```diff
@@ -1419,316 +1397,157 @@ def __init__(
-                        codes.cid7202.SourceImageForImageProcessingOperation
-                    derivation_src_img_item.PurposeOfReferenceCodeSequence = [
-                        CodedConcept.from_code(purpose_code)
-                    ]
-                    derivation_src_img_item.SpatialLocationsPreserved = 'YES'
-                    derivation_image_item.SourceImageSequence = [
-                        derivation_src_img_item,
-                    ]
-                    pffp_item.DerivationImageSequence.append(
-                        derivation_image_item
-                    )
                 else:
-                    logger.warning('spatial locations not preserved')
+                    dimension_index_values = []
+                pffg_item = self._get_pffg_item(
+                    segment_number=segment_number,
+                    dimension_index_values=dimension_index_values,
+                    plane_position=plane_positions[plane_index],
+                    source_images=source_images,
+                    source_image_index=plane_index,
+                    are_spatial_locations_preserved=are_spatial_locations_preserved,  # noqa: E501
+                    has_ref_frame_uid=has_ref_frame_uid,
+                    coordinate_system=self._coordinate_system,
+                )
+                pffg_sequence.append(pffg_item)
 
-                identification = Dataset()
-                identification.ReferencedSegmentNumber = int(segment_number)
-                pffp_item.SegmentIdentificationSequence = [
-                    identification,
-                ]
-                self.PerFrameFunctionalGroupsSequence.append(pffp_item)
-                self.NumberOfFrames += 1
-
-            if is_encaps:
-                # Encode this frame and add to the list for encapsulation
-                # at the end
-                for f in contained_plane_index:
-                    full_frames_list.append(self._encode_pixels(planes[f]))
-            else:
-                # Concatenate the 1D array for re-encoding at the end
-                full_pixel_array = np.concatenate([
-                    full_pixel_array,
-                    planes[contained_plane_index].flatten()
-                ])
+                # Add the segmentation pixel array for this frame to the list
+                if is_encaps:
+                    # Encode this frame and add to the list for encapsulation
+                    # at the end
+                    full_frames_list.append(segment_array[plane_index])
+                else:
+                    # Concatenate the 1D array for re-encoding at the end
+                    full_frames_list.append(
+                        segment_array[plane_index].flatten()
+                    )
 
-            self.SegmentSequence.append(segment_descriptions[i])
+        self.PerFrameFunctionalGroupsSequence = pffg_sequence
+        self.NumberOfFrames = len(pffg_sequence)
 
         if is_encaps:
```

Comment:
This part could potentially go into a separate `_add_pixel_data` helper method at some point:

```python
def _add_pixel_data(self, frames: List[numpy.ndarray], is_encapsulated: bool) -> None
```

Reply:
I actually rewrite this piece substantially in an upcoming pull request related to multiprocessing so I'm going to leave as is for now. Agreed that it doesn't make a huge amount of sense at the moment